rascal2/dialogs/project_dialog.py

```diff
@@ -0,0 +1,209 @@
+
+        self.create_button = QtWidgets.QPushButton(" Create", self)
+        self.create_button.setIcon(QtGui.QIcon(path_for("create.png")))
+        self.create_button.clicked.connect(self.create_project)
+        self.create_button.setStyleSheet(self._button_style.format("#0D69BB"))
+
+        self.cancel_button = QtWidgets.QPushButton(" Cancel", self)
+        self.cancel_button.setIcon(QtGui.QIcon(path_for("cancel.png")))
+        self.cancel_button.clicked.connect(self.cancel_project_creation)
+        self.cancel_button.setStyleSheet(self._button_style.format("#E34234"))
+
+    def cancel_project_creation(self) -> None:
+        """
+        Cancel project creation.
+        """
+        self.close() if self.parent().new_project_action_called else self.parent().toggleView()
+        self.reset_variables()
+
+    def reset_variables(self) -> None:
+        """
+        Reset variables.
+        """
+        self.project_folder.setText("")
+        self.project_name.setText("")
+        self.name_error = False
+        self.folder_error = False
+
+    def open_folder_selector(self) -> None:
+        """
+        Open folder selector.
+        """
+        folder_path = QtWidgets.QFileDialog.getExistingDirectory(self, "Select Folder")
+        if folder_path:
+            files_in_folder = list(filter(lambda x: x[0] != ".", os.listdir(folder_path)))
+            if files_in_folder:
+                self.handle_error("folder")
+            else:
+                self.handle_success("folder")
+                self.project_folder.setText(folder_path)
+
+    def handle_error(self, tag: Literal["name", "folder"]) -> None:
+        """
+        Display error msgs.
+
+        Parameters
+        ----------
+        tag: Literal["name", "folder"]
+            Specifies whether to show the error
+            for the project name or folder.
+        """
+        getattr(self, "project_" + tag).setStyleSheet(self._line_edit_error_style)
+        getattr(self, "project_" + tag + "_error").show()
+        setattr(self, tag + "_error", True)
+
+    def handle_success(self, tag: Literal["name", "folder"]) -> None:
+        """
+        Hide error msgs.
+
+        Parameters
+        ----------
+        tag: Literal["name", "folder"]
+            Specifies whether to clear the error
+            for the project name or folder.
+        """
+        getattr(self, "project_" + tag).setStyleSheet("")
+        getattr(self, "project_" + tag + "_error").hide()
+        setattr(self, tag + "_error", False)
+
+    def verify_inputs(self, tag: Literal["name", "folder"]) -> None:
+        """
+        Verify inputs.
+
+        Parameters
+        ----------
+        tag: Literal["name", "folder"]
+            Specifies whether the input verification
+            is for the project name or folder.
+        """
+        if getattr(self, "project_" + tag).text() == "":
+            self.handle_error(tag)
+        else:
+            self.handle_success(tag)
```

Comment:
the stuff here with `getattr` and the strings is very cool but quite hard to understand what's going on - also makes it harder to see where `self.project_name` and `self.project_folder` is being edited and read (because I can't use ctrl+f or 'find references' tools in IDEs to find these)

this is partially because of how nested the functions are - it would be much clearer to either call the `getattr` and `setattr` directly in `verify_inputs` rather than making them their own function, each of which are only called once

or even further, you may disagree depending on how much you care about DRY but you could just not bother with the abstraction:

```python
def verify_name():
    """
    Verify the project name.
    """
    if project_name.text() == "":
        project_name.setStyleSheet("")
        project_name_error.hide()
        return False
    else:
        # etc...

def verify_folder():
# as above...
```
and then it's immediately clear what is being verified where. and what attributes are being edited.

Reply:
I started with this and noticed that the functions were very similar and refactored them.